controlplane/internal/dal/postgres.go

```diff
@@ -458,6 +458,39 @@ func (d *Postgres) InsertMetricEntry(ctx context.Context, metric Metric) error {
 	})))
 }
 
+func (d *Postgres) GetMetricsForSourceModules(ctx context.Context, modules []string) ([]Metric, error) {
+	dbMetrics, err := d.db.GetMetricsBySourceModules(ctx, modules)
+	if err != nil {
+		return []Metric{}, errors.Wrap(translatePGError(err), "could not get metrics")
+	}
+	metrics, err := slices.MapErr(dbMetrics, func(in sql.Metric) (Metric, error) {
+		var datapoint DataPoint
+		if err := json.Unmarshal((in.Value), &datapoint); err != nil {
+			return Metric{}, errors.Wrapf(err, "%q: could not unmarshal datapoint", in.Value)
```

Comment:
In general I find (personally, YMMV) it's better to log the database row ID rather than values - values can be very large and unreadable, and can potentially leak PII. Also without the row ID it's difficult to track down where the corrupt data is.

It's a little more hassle to debug from a row ID but there's zero chance of either of those.